### ISSUE TITLE
Calico Issue Unable to autoassign an address - pools are likely exhausted

### ISSUE THREAD

**Issue by vamshisiddarth** [projectcalico/cni-plugin#920]
## Your Environment

* CNI plugin version: v3.15.1
* Orchestrator version: Kubernetes
* Operating System and version: Fedora CoreOS 32.20200629.3.0

**Description**

We have seen some strange issue today on one of our environments where the Calico pods were going into "CrashLoopBackOff" state on the new spot instances/nodes which were created as the old instances were terminated. As a result, all the pods which have been scheduled to run on these nodes are going into "ContainerCreating" state and stuck forever. We have checked the calico pods logs and the kubelet logs on the nodes which had issues.

###### Note: I replaced the IP's, Node Names and Application names in these logs.

Calico Pod Logs
```log
$ kubectl logs -f calico-node-w5p8l -n kube-system
2020-07-27 12:32:20.818 [INFO][10] startup/startup.go 299: Early log level set to info
2020-07-27 12:32:20.818 [INFO][10] startup/startup.go 315: Using NODENAME environment for node name
2020-07-27 12:32:20.818 [INFO][10] startup/startup.go 327: Determined node name: <Node-Name>
2020-07-27 12:32:20.819 [INFO][10] startup/startup.go 359: Checking datastore connection
2020-07-27 12:32:20.924 [INFO][10] startup/startup.go 383: Datastore connection verified
2020-07-27 12:32:20.925 [INFO][10] startup/startup.go 104: Datastore is ready
2020-07-27 12:32:21.225 [INFO][10] startup/startup.go 664: Using autodetected IPv4 address on interface ens5: <CIDR-block>
2020-07-27 12:32:21.225 [INFO][10] startup/startup.go 727: No AS number configured on node resource, using global value
2020-07-27 12:32:21.225 [INFO][10] startup/startup.go 201: Setting NetworkUnavailable to False
2020-07-27 12:32:21.421 [INFO][10] startup/startup.go 761: Skipping IP pool configuration
2020-07-27 12:32:21.758 [INFO][10] startup/startup.go 234: Using node name: <Node-Name>
2020-07-27 12:32:22.412 [INFO][32] tunnel-ip-allocator/ipam.go 1173: Releasing all IPs with handle 'wireguard-tunnel-addr-<Node-Name>'
2020-07-27 12:32:22.796 [INFO][32] tunnel-ip-allocator/allocateip.go 267: Assign a new tunnel address type="ipipTunnelAddress"
2020-07-27 12:32:22.796 [INFO][32] tunnel-ip-allocator/allocateip.go 343: Release any old tunnel addresses IP="" type="ipipTunnelAddress"
2020-07-27 12:32:22.796 [INFO][32] tunnel-ip-allocator/ipam.go 1173: Releasing all IPs with handle 'ipip-tunnel-addr-<Node-Name>'
2020-07-27 12:32:22.884 [INFO][32] tunnel-ip-allocator/allocateip.go 354: Assign new tunnel address IP="" type="ipipTunnelAddress"
2020-07-27 12:32:22.884 [INFO][32] tunnel-ip-allocator/ipam.go 88: Auto-assign 1 ipv4, 0 ipv6 addrs for host '<Node-Name>'
2020-07-27 12:32:22.954 [INFO][32] tunnel-ip-allocator/ipam.go 314: Looking up existing affinities for host handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.006 [INFO][32] tunnel-ip-allocator/ipam.go 372: Ran out of existing affine blocks for host handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.115 [INFO][32] tunnel-ip-allocator/ipam.go 437: No more affine blocks, but need to allocate 1 more addresses - allocate another block handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.115 [INFO][32] tunnel-ip-allocator/ipam.go 441: Looking for an unclaimed block handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.306 [INFO][32] tunnel-ip-allocator/ipam.go 446: No free blocks available for allocation handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.306 [INFO][32] tunnel-ip-allocator/ipam.go 524: Attempting to assign 1 more addresses from non-affine blocks handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.306 [INFO][32] tunnel-ip-allocator/ipam.go 528: Looking for blocks with free IP addresses handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.485 [INFO][32] tunnel-ip-allocator/ipam.go 548: Attempting to assign IPs from non-affine block <CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.485 [INFO][32] tunnel-ip-allocator/ipam.go 811: Attempting to assign 1 addresses from block block=<CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.485 [INFO][32] tunnel-ip-allocator/ipam.go 822: Block is full block=<CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [INFO][32] tunnel-ip-allocator/ipam.go 548: Attempting to assign IPs from non-affine block <CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [INFO][32] tunnel-ip-allocator/ipam.go 811: Attempting to assign 1 addresses from block block=<CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [INFO][32] tunnel-ip-allocator/ipam.go 822: Block is full block=<CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [WARNING][32] tunnel-ip-allocator/ipam.go 536: All addresses exhausted in pool <CIDR-block> handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [INFO][32] tunnel-ip-allocator/ipam.go 570: Auto-assigned 0 out of 1 IPv4s: [] handle="ipip-tunnel-addr-<Node-Name>" host="<Node-Name>"
2020-07-27 12:32:23.583 [FATAL][32] tunnel-ip-allocator/allocateip.go 425: Unable to autoassign an address - pools are likely exhausted. type="ipipTunnelAddress"
Calico node failed to start
```

## Expected Behavior
Calico should be able to allocate the IP address for the pods.

We are using [Typhoon](https://github.com/poseidon/typhoon) module for  Kubernetes cluster creation. Please let us know if you are aware of any such issues and also the way to fix it, if any. Let me know if you need any further details from me.

**Comment by caseydavenport** [comment]
@vamshisiddarth from the logs, it appears like Calico is attempting to allocate a new address for the node but is unable to because there are no free addresses left.

This could be a result of just running too many workloads / nodes for the given IP pools in the system, but more than likely it's a result of an IP address leak somewhere in the system, resulting in taken addresses not being released back into the available pool.

A few questions that would help diagnose here:
- How do you add / remove nodes from the cluster? Do you drain those nodes of pods before shutting them down?
- What version of Calico are you using? 

Further diagnosis will probably require one or more of the following:

- IPAM state of the cluster - e.g., `kubectl get ipamblocks -o yaml`
- Logs from the Calico kube-controllers pod, which is responsible for garbage collecting addresses from deleted nodes.

**Comment by vamshisiddarth** [comment]
Hey Casey, we were running our kubernetes cluster on the aws spot instances and the issue happened when there was too many instances terminated because of the cost difference. There were too many new nodes adding. We were deleting all the old nodes using the kubectl delete node. 
We were directly deleting the nodes instead of draining them first. I just released after reading your question that this was an available option.
This cluster was like 10 days old with around 100 pods running on it including the pods in kube-system as well.

We stopped seeing this issue since there is no spot instances being deleted that frequently. Will keep an eye and let you if this happens again. Thank you.
